File: jmlparser-jml-tools/src/main/java/com/github/jmlparser/smt/model/SExpr.java
package com.github.jmlparser.smt.model;

import com.github.javaparser.resolution.types.ResolvedType;
import com.github.jmlparser.smt.solver.AppendableTo;
import org.jetbrains.annotations.Nullable;

import java.io.PrintWriter;
import java.io.StringWriter;

/**
 * @author Alexander Weigl
 * @version 1 (07.08.22)
 */
public abstract class SExpr implements AppendableTo {
    @Nullable
    ResolvedType javaType;
    @Nullable
    SmtType smtType;

    @Nullable
    public ResolvedType getJavaType() {
        return javaType;
    }

    @Nullable
    public SmtType getSmtType() {
        return smtType;
    }

    public SList asList() {
        return (SList) this;
    }

    public String asSymbolValue() {
        return ((SAtom) this).getValue();
    }

    @Override
    public String toString() {
        StringWriter sw = new StringWriter();
        appendTo(new PrintWriter(sw));
        return sw.toString();
    }

    /*public int asNumberValue() {
        return ((com.github.jmlparser.smt.SExpr.SNumber) this).value;
    }*/
}

File: tools/smt/src/main/kotlin/io/github/jmltoolkit/smt/model/SExpr.kt
package io.github.jmltoolkit.smt.model

import com.github.javaparser.resolution.types.ResolvedType
import io.github.jmltoolkit.smt.solver.AppendableTo
import java.io.PrintWriter
import java.io.StringWriter

/**
 * @author Alexander Weigl
 * @version 1 (07.08.22)
 */
abstract class SExpr(
    var javaType: ResolvedType? = null,
    var smtType: SmtType? = null
) : AppendableTo {

    fun asList(): SList {
        return this as SList
    }

    fun asSymbolValue(): String {
        return (this as SAtom).value
    }

    override fun toString(): String {
        val sw = StringWriter()
        appendTo(PrintWriter(sw))
        return sw.toString()
    }
}
